src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java

```diff
@@ -0,0 +1,241 @@
+  /**
+   * For a single value Object, convert to the type defined in the metadata
+   */
+  private static Object convertSingleObjectToDefinedType(Object value, VCFHeaderLineType type) {
+    // values and their types have already been parsed by VCFCodec.decode()
+    if (!(value instanceof String)) {
+      return value;
+    }
+
+    String valueStr = (String)value;
+    valueStr = valueStr.trim();   // for cases like " 27", ignore the space in list element
+
+    if (valueStr.equals(MISSING_FIELD_VALUE)) {
+      return null;
+    }
+
+    // double check if the String value match the type and also is a number
+    boolean isNumber = valueStr.matches("[-+]?\\d+(\\.\\d+)?");
+
+    if (type == VCFHeaderLineType.Integer && isNumber) {
+      return Integer.parseInt(valueStr);
+    } else if (type == VCFHeaderLineType.Float && isNumber) {
+      return Double.parseDouble(valueStr);
+    } else {
+      // default String value
+      return value;
+    }
+  }
+
+  private static void addGenotypes(TableRow row, List<Allele> alleles,
+                                   Map<String, Integer> alleleIndexingMap) {
+    List<Integer> genotypes = new ArrayList<>();
+    for (Allele allele : alleles) {
+      String alleleStr = allele.getDisplayString();
+      if (alleleStr.equals(MISSING_FIELD_VALUE)) {
+        genotypes.add(-1);
+      } else {
+        genotypes.add(alleleIndexingMap.getOrDefault(alleleStr, -1));
+      }
+    }
+    row.set(ColumnKeyConstants.CALLS_GENOTYPE, genotypes);
+  }
+
+  private static void addInfoAndPhaseSet(TableRow row, Genotype genotype, VCFHeader vcfHeader) {
+    String phaseSet = "";
+
+    if (genotype.hasAD()) { row.set("AD", genotype.getAD()); }
+    if (genotype.hasDP()) { row.set("DP", genotype.getDP()); }
+    if (genotype.hasGQ()) { row.set("GQ", genotype.getGQ()); }
+    if (genotype.hasPL()) { row.set("PL", genotype.getPL()); }
+    Map<String, Object> extendedAttributes = genotype.getExtendedAttributes();
+    for (String extendedAttr : extendedAttributes.keySet()) {
+      if (extendedAttr.equals(PHASESET_FORMAT_KEY)) {
+        String phaseSetValue = extendedAttributes.get(PHASESET_FORMAT_KEY).toString();
+        phaseSet = dealWithMissingFieldValue(phaseSetValue);
+      } else {
+        // The rest of fields need to be converted to the right type by VCFCodec decode function
+        row.set(extendedAttr, convertToDefinedType(extendedAttributes.get(extendedAttr),
+                vcfHeader.getFormatHeaderLine(extendedAttr).getType()));
+      }
+    }
+    if (phaseSet == null || phaseSet.length() != 0) {
+      // phaseSet is presented(MISSING_FIELD_VALUE('.') or real value)
+      row.set(ColumnKeyConstants.CALLS_PHASESET, phaseSet);
+    } else {
+      row.set(ColumnKeyConstants.CALLS_PHASESET, DEFAULT_PHASESET);
+    }
+  }
+
+
+  /**
+   * In some fields that values are MISSING_FIELD_VALUE('.'), should set null in BQ row.
+   */
+  private static String dealWithMissingFieldValue(String value) {
```

Comment:
if we choose to keep this function, what about calling it handleMissingFieldValue(...)?

Reply:
I will delete this method, looks useless to separate this into one function